### ISSUE TITLE
db connection closed and worker hangs with celery 4.2+

### ISSUE THREAD

**Issue by craigds** [celery/celery#4878]
This might be two separate bugs, but they might be related.

## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
- [x] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce

Run multiple celery workers with `celery multi`, each with its own queue, then pass tasks to them.

## Expected behavior

tasks should run on all workers.

### Most workers explode:

Most nodes give this, for every task:

```
[2018-07-04 02:52:21,280: WARNING/ForkPoolWorker-1] Request: <Context: {'origin': 'gen19542@vm', u'args': [], u'chain': None, 'root_id': '0c008f5a-507d-4b08-8082-0dcc0ab3495d', 'expires': None, u'is_eager': False, u'correlation_id': '0c008f5a-507d-4b08-8082-0dcc0ab3495d', u'chord': None, u'reply_to': '5e48e3d9-0b67-3122-b0c5-9b83efd9e5ca', 'id': '0c008f5a-507d-4b08-8082-0dcc0ab3495d', 'kwargsrepr': '{}', 'lang': 'py', 'retries': 0, 'task': 'dandelion.geostore.celery.debug_task', 'group': None, 'timelimit': [None, None], u'delivery_info': {u'priority': 0, u'redelivered': False, u'routing_key': u'celery', u'exchange': u''}, u'hostname': u'celery@vm', 'called_directly': False, 'parent_id': None, 'argsrepr': '()', u'errbacks': None, u'callbacks': None, u'kwargs': {}, 'eta': None, '_protected': 1}>
[2018-07-04 02:52:21,288: WARNING/ForkPoolWorker-1] /kx/dandelion/venv/local/lib/python2.7/site-packages/celery/app/trace.py:561: RuntimeWarning: Exception raised outside body: OperationalError('could not receive data from server: Bad file descriptor\n',):
Traceback (most recent call last):
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/celery/app/trace.py", line 442, in trace_task
    uuid, retval, task_request, publish_result,
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/celery/backends/base.py", line 144, in mark_as_done
    self.store_result(task_id, result, state, request=request)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/celery/backends/base.py", line 316, in store_result
    request=request, **kwargs)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django_celery_results/backends/database.py", line 29, in _store_result
    meta=meta,
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django_celery_results/managers.py", line 50, in _inner
    return fun(*args, **kwargs)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django_celery_results/managers.py", line 119, in store_result
    obj, created = self.get_or_create(task_id=task_id, defaults=fields)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 473, in get_or_create
    return self.get(**lookup), False
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 379, in get
    num = len(clone)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 238, in __len__
    self._fetch_all()
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 1087, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/query.py", line 54, in __iter__
    results = compiler.execute_sql()
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 835, in execute_sql
    cursor.execute(sql, params)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/raven/contrib/django/client.py", line 127, in execute
    return real_execute(self, sql, params)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/kx/dandelion/venv/local/lib/python2.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
OperationalError: could not receive data from server: Bad file descriptor
```

I guess something's closing the django db connection?

### One worker succeeds but then hangs

In another worker (`iris`), one task actually _succeeds_ without hitting the above message, except that then the worker hangs forever. Telling it to stop with `stopwait` command just hangs the stopwait command too. There is nothing in the log for this worker from the time the task completes (the task completion is also not logged, but if I insert print statements at the end of the task they end up in the log)

Our init script to run all the workers is here: https://gist.github.com/craigds/42f3ad627d68e94aefa909c2b1a4a0c0

As far as I can tell, the tasks aren't doing anything particularly interesting. Putting `pass` as the task body doesn't change anything.

Tasks are defined like this:

```python
from celery import shared_task

@shared_task(bind=True, queue='iris')
def queue_slow_rendered_tiles(self):
...
```

This works well with celery 4.1.1. What other information can I provide that might illuminate the issue?

# Theory

The only difference I can think of with the `iris` worker is that it doesn't use the django database in its tasks. Presumably the other workers, that _do_ use the database, are exploding prior to this hang, whereas the `iris` one is skipping the db connection close and hitting a different bug. Maybe thats a clue?

# celery report output

I grepped the relevant bits, the entirety of our django settings is not relevant here:

```
software -> celery:4.2.0 (windowlicker) kombu:4.2.1 py:2.7.6
            billiard:3.5.0.3 py-amqp:2.3.2
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:django-db


CELERY_BROKER_URL: u'amqp://appname:********@localhost:5672/appname'
CELERY_TASK_QUEUES:
    (<unbound Queue bcast.e39987e0-37c3-4eb7-b63a-dace642ba7da -> <unbound Exchange notification_broadcast(fanout)> -> >,)
CELERY_BEAT_SCHEDULER: 'django_celery_beat.schedulers:DatabaseScheduler'
CELERY_IMPORTS:
    ('appname', 'otherappname')
CELERY_RESULT_BACKEND: 'django-db'
CELERY_RESULT_EXPIRES: 3600

```

These seem relevant too: 
```
celery==4.2.0
django-celery-beat==1.1.1
django-celery-results==1.0.1
kombu==4.2.1
amqp==2.3.2
```

**Comment by canassa** [comment]
I also had a similar issue after upgrading to 4.2. We noticed that everytime that we restarted the workers, some of them would randomly "hang". Rabbitmq would still report that a worker was connected to it but the worker wouldn't process any task until restarted again.

**Comment by jarhainx** [comment]
I have a similar issue.
My config
```
software -> celery:4.2.0 (windowlicker) kombu:4.2.1 py:3.5.3
            billiard:3.5.0.3 redis:2.10.6
platform -> system:Linux arch:64bit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis://192.168.1.151:6379/13

result_serializer: 'json'
broker_url: 'redis://192.168.1.151:6379/13'
broker_transport_options: {
 'visibility_timeout': 43200}
imports: ['ocpp_server.v16.tasks']
timezone: 'Europe/Paris'
beat_schedule: {.......}
result_backend: 'redis://192.168.1.151:6379/13'
accept_content: ['json']
settings: <LazySettings "project.settings">
task_serializer: 'json'
```
Celery is installed on 9.4 debian LXC Container managed by Proxmox 5.2-5

When I rollback to 4.1 everything is ok!

**Comment by auvipy** [comment]
If roling back to 4.1 fix it then this should be a regression in new release! could any of you help to investigate the issue?

**Comment by andialbrecht** [comment]
I see the same when switching from 4.1.0 to 4.2.1. 

@auvipy could you give me a pointer where I could start to dig down this issue?

**Comment by wrvdklooster** [comment]
We saw something similar but only after moving also to `django_celery_beat.schedulers:DatabaseScheduler`. When using `PersistentScheduler` the issue is gone. When going back to 4.1.1 the issue is also gone.

When using DatabaseScheduler and 4.2.1 the issue is there.

**Comment by mumumumu** [comment]
I'm also seeing a similar issue on 4.2.1 using `DatabaseScheduler`. ~10% of tasks were never making it to the workers. Rolling back to 4.1.0 fixed it.

**Comment by marcomayer** [comment]
+1 having the same when restarting workers. One job per worker throws the expection about 25% of the time, next one would work again.

**Comment by auvipy** [comment]
can anyone bisect the commit?

**Comment by godlark** [comment]
I can confirm that the issue exist. I faced this after upgrading celery from 4.1.0 to 4.2.0. I use `celer[redis]` on Heroku. I run only one worker with `celery worker --app=my_app`. I think it's the simplest configuration possible. Configuration for celery in django settings.py:
```
BROKER_URL = env_as_str('BROKER_URL')
CELERY_RESULT_BACKEND = BROKER_URL
CELERY_ACCEPT_CONTENT = ['json']
CELERY_TASK_SERIALIZER = 'json'
CELERY_TASK_IGNORE_RESULT = True
CELERY_ALWAYS_EAGER = env_as_bool('CELERY_ALWAYS_EAGER', False)
broker_pool_limit = env_as_int('BROKER_POOL_LIMIT', 0)
redis_max_connections = env_as_int('CELERY_REDIS_MAX_CONNECTIONS', 40)
```

**Comment by godlark** [comment]
I can confirm that the issue exist. I faced this after upgrading celery from 4.1.0 to 4.2.0. I use `celer[redis]` on Heroku. I run only one worker with `celery worker --app=my_app`. I think it's the simplest configuration possible. Configuration for celery in django settings.py:
```
BROKER_URL = env_as_str('BROKER_URL')
CELERY_RESULT_BACKEND = BROKER_URL
CELERY_ACCEPT_CONTENT = ['json']
CELERY_TASK_SERIALIZER = 'json'
CELERY_TASK_IGNORE_RESULT = True
CELERY_ALWAYS_EAGER = env_as_bool('CELERY_ALWAYS_EAGER', False)
broker_pool_limit = env_as_int('BROKER_POOL_LIMIT', 0)
redis_max_connections = env_as_int('CELERY_REDIS_MAX_CONNECTIONS', 40)
```

The encontered error

The error is:
```
 Nov 22 11:28:18 xxx app[worker] info [2018-11-22 10:28:18,007: ERROR/ForkPoolWorker-3] Task yyy.zzz.views.vvv.foo_bar[e57f6154-95f2-4f21-b2c6-056b6a99c189] raised unexpected: OperationalError('SSL SYSCALL error: Bad file descriptor\n',)
Nov 22 11:28:18 xxx app[worker] info Traceback (most recent call last):
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Nov 22 11:28:18 xxx app[worker] info     return self.cursor.execute(sql, params)
Nov 22 11:28:18 xxx app[worker] error psycopg2.OperationalError: SSL SYSCALL error: Bad file descriptor
Nov 22 11:28:18 xxx app[worker] info
Nov 22 11:28:18 xxx app[worker] info
Nov 22 11:28:18 xxx app[worker] info The above exception was the direct cause of the following exception:
Nov 22 11:28:18 xxx app[worker] info
Nov 22 11:28:18 xxx app[worker] info Traceback (most recent call last):
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/celery/app/trace.py", line 382, in trace_task
Nov 22 11:28:18 xxx app[worker] info     R = retval = fun(*args, **kwargs)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/celery/app/trace.py", line 641, in __protected_call__
Nov 22 11:28:18 xxx app[worker] info     return self.run(*args, **kwargs)
Nov 22 11:28:18 xxx app[worker] info   File "/app/yyy/zzz/views/vvv.py", line 44, in foo_bar
Nov 22 11:28:18 xxx app[worker] info     end_date
Nov 22 11:28:18 xxx app[worker] info   File "/app/yyy/zzz/uuu.py", line 40, in bar_foo
Nov 22 11:28:18 xxx app[worker] info     abc__has_exception=True,
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/query.py", line 387, in count
Nov 22 11:28:18 xxx app[worker] info     return self.query.get_count(using=self.db)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/sql/query.py", line 491, in get_count
Nov 22 11:28:18 xxx app[worker] info     number = obj.get_aggregation(using, ['__count'])['__count']
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/sql/query.py", line 476, in get_aggregation
Nov 22 11:28:18 xxx app[worker] info     result = compiler.execute_sql(SINGLE)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1068, in execute_sql
Nov 22 11:28:18 xxx app[worker] info     cursor.execute(sql, params)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/raven/contrib/django/client.py", line 123, in execute
Nov 22 11:28:18 xxx app[worker] info     return real_execute(self, sql, params)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/backends/utils.py", line 68, in execute
Nov 22 11:28:18 xxx app[worker] info     return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/backends/utils.py", line 77, in _execute_with_wrappers
Nov 22 11:28:18 xxx app[worker] info     return executor(sql, params, many, context)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Nov 22 11:28:18 xxx app[worker] info     return self.cursor.execute(sql, params)
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/utils.py", line 89, in __exit__
Nov 22 11:28:18 xxx app[worker] info     raise dj_exc_value.with_traceback(traceback) from exc_value
Nov 22 11:28:18 xxx app[worker] info   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/backends/utils.py", line 85, in _execute
Nov 22 11:28:18 xxx app[worker] info     return self.cursor.execute(sql, params)
Nov 22 11:28:18 xxx app[worker] error django.db.utils.OperationalError: SSL SYSCALL error: Bad file descriptor
Nov 22 11:28:18 xxx app[worker] info
Nov 22 11:28:20 xxx app[worker] info [2018-11-22 10:28:19,883: ERROR/ForkPoolWorker-5] Task yyy.kpi.views.ttt[6871733a-dfb8-48f1-b68d-fff5c2182c4d] raised unexpected: OperationalError('SSL SYSCALL error: Bad file descriptor\n',) 
```

**Comment by Alan-Solitar** [comment]
> SSL SYSCALL error: Bad file descriptor

Had the same exact issue yesterday when I tried to upgrade my celery to 4.2.1.

I am using 'django_celery_beat.schedulers:DatabaseScheduler' but it seems celery is unable to read the schedule from my postgres database.

Just as many others reported, downgrading to 4.1 fixed the issue.

**Comment by georgepsarakis** [comment]
As @colinrieger [pointed out](https://github.com/celery/celery/issues/4878#issuecomment-440063737), Celery takes into account the Django [CONN_MAX_AGE](https://docs.djangoproject.com/en/2.2/ref/settings/#conn-max-age) setting. So you will need to set it to zero, for the workers specifically, so that you revert to the previous behaviour.

**Comment by rixx** [comment]
Will there be a fix for this in 4.3? If this is intended behaviour (is it?), will there be instructions regarding this issues in the changelog/release notes?

**Comment by amureki** [comment]
I am also wondering what is the best way of having `CONN_MAX_AGE` for celery set separately apart from other workers?

**Comment by thedrow** [comment]
@georgepsarakis Any chance you'll be able to provide a fix/documentation changes for this issue?

**Comment by denizdogan** [comment]
I would also like to know how to properly override this specific setting for workers only. I use multiple settings files for my project, one for each environment and I'd hate to write one worker-specific setting for each environment. It's far from obvious how to go about this in my opinion.

**Comment by Chronial** [comment]
I stumbled upon the root cause of this: #5483

**Comment by mlissner** [comment]
Oof. Between this bug, which requires version 4.1, and bug https://github.com/celery/celery/issues/3808, which requires > 4.1, we're struggling to find a version of Celery that *works*. I guess our only solution is to upgrade *back* to 4.3 (we downgraded to fix this issue previously before discovering #3808), and then set MAX_CONN_AGE to 0. 

Could this at least be documented as a known bug or something? The docs on the Django page could just say:

## Known Issues

CONN_MAX_AGE other than zero is known to cause issues according to bug #4878. Until this is fixed, please set CONN_MAX_AGE to zero.

**Comment by auvipy** [comment]
> Oof. Between this bug, which requires version 4.1, and bug #3808, which requires > 4.1, we're struggling to find a version of Celery that _works_. I guess our only solution is to upgrade _back_ to 4.3 (we downgraded to fix this issue previously before discovering #3808), and then set MAX_CONN_AGE to 0.
> 
> Could this at least be documented as a known bug or something? The docs on the Django page could just say:
> 
> ## Known Issues
> CONN_MAX_AGE other than zero is known to cause issues according to bug #4878. Until this is fixed, please set CONN_MAX_AGE to zero.

I warmly welcome your PR with proposed Idea :)

**Comment by Chronial** [comment]
@auvipy As a quickfix, you could roll back #4292 – it doesn't provide anything useful at the moment. Before, celery didn't properly respect CONN_MAX_AGE > 0, now it crashes with CONN_MAX_AGE > 0. I would argue that the previous situation was strictly better.

**Comment by auvipy** [comment]
I am going to revert that for now

**Comment by auvipy** [comment]
Please review and check in your dev env

**Comment by mlissner** [comment]
PR for docs created in #5516.

**Comment by auvipy** [comment]
> PR for docs created in #5516.

Please check the previous conversations. we decided to revert back, you can share your inputs about that as well

**Comment by mlissner** [comment]
Sorry, I'm a bit out of my wheelhouse here and I've got a deadline today. To test can I just do:

$ pip install git+git://https://github.com/auvipy/celery.git@e5bfd368fa31d1cce1a756b91099465c5616e198

That seems to be the right thing from your PR: 

https://github.com/auvipy/celery/commit/e5bfd368fa31d1cce1a756b91099465c5616e198

**Comment by auvipy** [comment]
should have been fixed in masters

**Comment by Asday** [comment]
[Documentation](http://docs.celeryproject.org/en/latest/django/first-steps-with-django.html#known-issues) still says this is broken.

**Comment by paaksing** [comment]
As of now (may 2020) the docs still says this is broken

**Comment by alexandervlpl** [comment]
Any chance this fix will be backported to 4.3? I can't upgrade to 4.4 because of this [problem](https://groups.google.com/g/celery-users/c/LKXvgPslH6k) or 5.0 because it requires Python 3.
